File: DSA/Sauhard_Semalty_590013630/Day18Question3.py
import sys

def dutch_flag_sort(nums: list[int]) -> None:
    "Sorts list in place using dutch national flag sort."

    low = 0
    mid = 0
    high = len(nums) - 1

    while mid <= high:
        if nums[mid] == 0:
            nums[low], nums[mid] = nums[mid], nums[low]
            low += 1
            mid += 1
        elif nums[mid] == 1:
            mid += 1
        else:
            nums[high], nums[mid] = nums[mid], nums[high]
            high -= 1


def main() -> None:
    "The entry point of the program."

    try:
        n = int(input("Enter the size of the array: "))
    except ValueError:
        print("Invalid size!")
        sys.exit(-1)

    # Check if length is valid
    if n <= 0:
        print("Size must be greater 0.")
        sys.exit(-1)

    nums: list[int] = []

    print("+--Enter the values--+")
    for _ in range(n):
        try:
            num = int(input("> "))
        except ValueError:
            print("Invalid value!")
            sys.exit(-1)

        nums.append(num)

    dutch_flag_sort(nums)  # Sort list in place
    print(nums)


if __name__ == "__main__":
    main()

File: DSA/ShatakshiMishra_590015008/Day18Question3.cpp
#include <iostream>
#include <vector>
using namespace std;

int main() {
    int n;
    cin >> n;
    vector<int> nums(n);

    for (int i = 0; i < n; ++i) {
        cin >> nums[i];
    }

    int low = 0, mid = 0, high = n - 1;

    while (mid <= high) 
    {
        if (nums[mid] == 0) 
        {
            swap(nums[low], nums[mid]);
            low++;
            mid++;
        } 
        else if (nums[mid] == 1) 
        {
            mid++;
        } 
        else 
        {
            swap(nums[mid], nums[high]);
            high--;
        }
    }

    for (int i = 0; i < n; ++i) 
    {
        cout << nums[i] << " ";
    }
    cout << endl;
}
